src/repository.rs

```diff
@@ -312,83 +349,115 @@ where
     }
 }
 
-impl<D> Repository<D> for HttpRepository<D>
+impl<C, D> Repository<D> for HttpRepository<C, D>
 where
+    C: Connect + Sync + 'static,
     D: DataInterchange,
 {
-    type TargetRead = Response;
-
     /// This always returns `Err` as storing over HTTP is not yet supported.
-    fn store_metadata<M>(
-        &self,
-        _: &MetadataPath,
-        _: &MetadataVersion,
-        _: &SignedMetadata<D, M>,
-    ) -> Result<()>
+    fn store_metadata<'a, M>(
+        &'a self,
+        _: &'a MetadataPath,
+        _: &'a MetadataVersion,
+        _: &'a SignedMetadata<D, M>,
+    ) -> TufFuture<'a, Result<()>>
     where
-        M: Metadata,
+        M: Metadata + 'static,
     {
-        Err(Error::Opaque(
-            "Http repo store metadata not implemented".to_string(),
-        ))
+        Box::pinned(
+            async {
+                Err(Error::Opaque(
+                    "Http repo store metadata not implemented".to_string(),
+                ))
+            },
+        )
     }
 
-    fn fetch_metadata<M>(
-        &self,
-        meta_path: &MetadataPath,
-        version: &MetadataVersion,
-        max_size: &Option<usize>,
+    fn fetch_metadata<'a, M>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        max_size: &'a Option<usize>,
         min_bytes_per_second: u32,
-        hash_data: Option<(&HashAlgorithm, HashValue)>,
-    ) -> Result<SignedMetadata<D, M>>
+        hash_data: Option<(&'static HashAlgorithm, HashValue)>,
+    ) -> TufFuture<'a, Result<SignedMetadata<D, M>>>
     where
-        M: Metadata,
+        M: Metadata + 'static,
     {
-        Self::check::<M>(meta_path)?;
+        Box::pinned(
+            async move {
+                Self::check::<M>(meta_path)?;
+
+                let components = meta_path.components::<D>(&version);
+                let resp = await!(self.get(&self.metadata_prefix, &components))?;
 
-        let resp = self.get(&self.metadata_prefix, &meta_path.components::<D>(&version))?;
+                let stream = resp
+                    .into_body()
+                    .compat()
+                    .map_err(|err| io::Error::new(io::ErrorKind::Other, err));
+
+                let mut reader = SafeReader::new(
+                    IntoAsyncRead::new(stream),
+                    max_size.unwrap_or(::std::usize::MAX) as u64,
+                    min_bytes_per_second,
+                    hash_data,
+                )?;
 
-        let read = SafeReader::new(
-            resp,
-            max_size.unwrap_or(::std::usize::MAX) as u64,
-            min_bytes_per_second,
-            hash_data,
-        )?;
-        Ok(D::from_reader(read)?)
+                let mut buf = Vec::new();
+                await!(reader.read_to_end(&mut buf))?;
+
+                Ok(D::from_reader(buf.as_slice())?)
```

Comment:
This may be a larger change to save for the future, but should the `DataInterchange` trait be changed to work over `AsyncRead`/`AsyncWrite` rather than `Read`/`Write` so as to prevent the intermediate buffer here?

Reply:
We could, but `DataInterchange` really is just doing a serde_json deserialization, that only supports deserializing from `&[u8]` and `std::io::Read`. I can however at least add an api for `DataInterchange::from_slice`, which may be a little faster.

Reply:
There's actually a new [TAP](https://github.com/theupdateframework/taps/pull/106) filed last month that adds support for multiple transport formats.

We also might want to explore other formats for optimizing storage and transport overhead, so I think it's worth keeping around right now.

Reply:
Ok so are you still shooting for compatibility with DockerHub or Notary or whatever it was?